llvm-general/src/LLVM/General/Internal/Module.hs

```diff
@@ -236,6 +236,19 @@ getDataLayout m = do
   dlString <- decodeM =<< FFI.getDataLayout m
   either fail return . runExcept . parseDataLayout A.BigEndian $ dlString
 
+-- | Build an LLVM.General.'Module' from a
+-- LLVM.General.AST.'LLVM.General.AST.Module' with a 'DataLayout'
+-- appropriate for the given 'TargetMachine'. This is important for
+-- native code generation as it impacts name mangling.
+withModuleFromASTForTargetMachine :: Context -> TargetMachine -> A.Module
```

Comment:
Another possiblity would be to set the datalayout on functions that care about it like I’ve done in the [orcjit](https://github.com/bscarlet/llvm-general/blob/llvm-3.9/llvm-general/src/LLVM/General/Internal/FFI/OrcJITC.cpp#L86) bindings (I stole that trick from the LLVM guys).

@acowley Do you recall which function needed the datalayout set and do we have that datalayout available when calling it?